director/projects/api/views/snapshots.py

```diff
@@ -0,0 +1,258 @@
+                    tag, pk
+                )
+            else:
+                error_message = str(e)
+            event.message = error_message
+            event.save()
+            raise ConflictError(error_message)
+        except Exception as e:
+            event.level = ProjectEventLevel.ERROR.value
+            event.success = False
+            event.message = str(e)
+            event.save()
+            raise
+        finally:
+            event.finished = timezone.now()
+            event.save()
+
+    @swagger_auto_schema(responses={status.HTTP_200_OK: SnapshotSerializer},)
+    def retrieve(self, request: Request, pk: int, number: int) -> Response:
+        """
+        Retrieve a snapshot of the project.
+
+        Returns details of the snapshot, including date/time `created`
+        and `completed`, and any `tag`.
+        """
+        # Check that the user has VIEW permissions for the project
+        self.get_project(request.user, pk=pk)
+        if not self.has_permission(ProjectPermissionType.VIEW):
+            raise PermissionDenied
+
+        snapshot = get_object_or_404(Snapshot, project=pk, version_number=number)
+        serializer = SnapshotSerializer(snapshot)
+        return Response(serializer.data)
+
+    @action(detail=True, methods=["get"], url_path="(?P<path>.+)")
+    def retrieve_file(
+        self, request: Request, pk: int, number: int, path: str
+    ) -> Response:
+        """
+        Retrieve a file within the snapshot of the project.
+
+        Returns the content of the file, either as is, or converted
+        to the desired format as specified in the `format` query parameter
+        or the `Accept` header.
+        """
+        # Check that the user has VIEW permissions for the project
+        project = self.get_project(request.user, pk=pk)
+        if not self.has_permission(ProjectPermissionType.VIEW):
+            raise PermissionDenied
+
+        snapshot = get_object_or_404(Snapshot, project=pk, version_number=number)
+
+        file_path = snapshot_path(snapshot, path)
+        if not os.path.exists(file_path) or os.path.isdir(file_path):
+            raise NotFound
+
+        # Determine the desired format
+        format = request.query_params.get("format")
+        if format is not None:
+            format_id = ConversionFormatId.from_id(format)
+        else:
+            accept = request.META.get("HTTP_ACCEPT")
+            if accept is not None and accept != "*/*":
+                media_types = [token.strip() for token in accept.split(",")]
+                format_id = ConversionFormatId.from_mimetype(media_types[0])
+            else:
+                format_id = None
+
+        raw = request.query_params.get("raw")
+        if raw or format_id is None:
+            # Respond with the file
+            response_path = file_path
+        else:
+            # Generate file for the format if necessary and return it
+            response_path = snapshot_path(
```

Comment:
This should be creating a `PublishedItem` instead of just putting the file in an arbitrary place.

Reply:
`PublishedItem` will need to be modified to store an output format to be queried on.

Reply:
PublishedItem could also contain the theme name so we know if we have generated one in a particular theme.

Reply:
I don't see the benefit of using `PublishedItem` here. Caching the file is just an optimization, and we might want to bust it e.g. if we make a fix to the HTML generated by Encoda. As we've discussed we need probably need to change our concept of what it means to "publish" something and move towards publishing an entire frozen snapshot (as a unit of reproducibility), submitting it to a data repository like Zenoda and getting a DOI. At that point, I think that the `PublishedItem` concept becomes useful as a prior-specification of which files in a project should get executed and converted (and to which format) just prior to publication. In contrast, this current functionality is just on-the-fly conversion with some caching thrown in for performance.